### DISCUSSION TITLE
🎉 Perseus v0.4.0 Goes Stable! 🎉

### DISCUSSION THREAD

**Discussion by arctic-hen7** [framesurge/perseus#270] [Announcements]
Today, after one year of beta versions, I am very happy to announce the stable release of Perseus v0.4.0, the latest major version of the [Perseus framework](https://framesurge.sh/perseus/en-US)! This version comes not only with a slew of new features and bug fixes, but also a *full rewrite* of the entire Perseus core, streamlining it to improve speed and reliability. Here are a few of the new features you might like:

- [Capsules](https://framesurge.sh/perseus/en-US/docs/next/capsules/intro), which allow you to split your pages into multiple *widgets*, all with access to Perseus' state generation platform, letting you delay loading parts of your pages trivially to improve load times, while encouraging code reuse and more idiomatic UI development
- An overhauled error management system, now with an intelligent distinction between different types of errors to make sure you never replace working content from the server with an error again
- Improved inbuilt optimizations through the CLI, making sure your app is fast from the second you run `perseus deploy`
- All state is now cached at the application-level, meaning users who go back to a page they've already visited will get there *instantly*, with no connection required (this works for capsules too, meaning pages that use the same widgets will load even faster)
- Upgraded to Sycamore v0.8 for faster reactive primitives and greatly improved developer ergonomics
- Added support for [custom API routes](https://github.com/framesurge/perseus/tree/main/examples/core/custom_server)
- Removed a *ton* of unnnecessary macros, and greatly improved the developer experience
- Made the CLI automatically install the tools it needs to build and optimize your app: `cargo install perseus-cli && perseus new my-app && cd my-app && perseus serve` *just works*, no dependencies needed (except Rust, we're not *that* good...yet)
- Added a more lightweight translator for inbuilt internationalization to trim down your app's bundle size, improving performance
- Added a panic handling system that lets you do crash analytics and handle critical failures gracefully in your app
- Removed old UI quirks to make styling *much* easier
- Improved hydration support to make your apps *even faster*, while fixing untold numbers of bugs
- Added `perseus check` to let your validate your app more quickly
- Added inbuilt minification of HTML/CSS/JS assets
- Made testing your app literally 2x more ergonomic
- Added server integrations for Axum and Rocket
- Fixed a heck of a lot of bugs!

To see the full list of changes, check out [the changelog](https://github.com/framesurge/perseus/blob/main/CHANGELOG.md). To learn how to migrate your existing Perseus app to v0.4.0, see [this page](https://framesurge.sh/perseus/en-US/docs/next/migrating).

If you're new to Perseus, [start here](https://framesurge.sh/perseus/en-US/docs/next/quickstart)!

Finally, I just want to thank the entire Perseus community, for helping us reach 1,300 stars (🤯), and for being so supportive and helpful during this (very long) development period, and throughout the life of the project. This release has a ton of features I've always wanted to see in a web framework, and we couldn't have built it without all your support. Thank you so much!

**Comment by seanpmyers** [comment]
🥳 This is awesome @arctic-hen7! 

Big thanks to you and all the contributors !

**Comment by menjaraz** [comment]
Is Perseus stale? No new release since then... while Rust is moving ahead. Last version doesn't work with new Cargo.

**Reply by arctic-hen7** [reply]
Somewhat stale at the moment unfortunately. You need to downgrade to an earlier version of Rust right now, we need to push some changes to fix that and a couple of other problems. I don't have the bandwidth to work on Perseus a great deal anymore, so welcome contributors! We've got a thread going at #336 on what needs to be done.
